nx/lib/nx/random.ex

```diff
@@ -0,0 +1,419 @@
+            impl(key, Nx.concatenate([Nx.iota(shape), Nx.iota(shape) + Nx.size(shape)]))
+            |> Nx.reshape({2, :auto})
+            |> Nx.as_type({:u, bit_width})
+
+          bits[0] <<< 32 ||| bits[1]
+
+        32 ->
+          impl(key, Nx.iota(shape))
+
+        _ ->
+          impl(key, Nx.iota(shape))
+          |> Nx.as_type({:u, bit_width})
+      end
+
+    Nx.reshape(bits, shape)
+  end
+
+  defnp impl(key, count) do
+    shape = Nx.shape(count)
+
+    reshaped_key = Nx.reshape(key, {2, 1})
+
+    reshaped_count =
+      Nx.reshape(count, {:auto})
+      |> Nx.as_type({:u, 32})
+
+    threefry2x32(reshaped_key, reshaped_count)
+    |> Nx.reshape(shape)
+    |> Nx.as_type({:u, 32})
+  end
+
+  # Check count
+  defn threefry2x32(key, count) do
+    padding =
+      Nx.size(count)
+      |> rem(2)
+
+    Nx.pad(count, 0, [{0, padding, 0}])
+    |> Nx.reshape({2, :auto})
+    |> Nx.as_type({:u, 32})
+    |> threefry2x32_20(key)
+    |> Nx.reshape({:auto})
+    |> Nx.pad(0, [{0, -padding, 0}])
+  end
+
+  defnp threefry2x32_20(xs, ks) do
+    rotations = Nx.tensor([[13, 15, 26, 6], [17, 29, 16, 24]], type: {:u, 8})
+    key1 = ks[0]
+    key2 = ks[1]
+
+    xs = Nx.add(ks, xs)
+
+    ks =
+      Nx.stack([
+        key2,
+        Nx.bitwise_xor(key1, key2)
+        |> Nx.bitwise_xor(0x1BD11BDA),
+        key1
+      ])
+      # we lose precision on purpose
+      |> Nx.as_type({:u, 32})
+
+    state = {xs, ks, rotations}
+
+    {_, {nxs, _, _}} =
+      while {x = 0, state}, Nx.less(x, 5) do
+        {x + 1, rolled_loop_step(x, state)}
+      end
+
+    nxs
+  end
+
+  defnp apply_round(xs, rot) do
+    y1 = Nx.sum(xs, axes: [0])
+
+    y2 =
+      rotate_left(xs[1], rot)
+      |> Nx.bitwise_xor(y1)
+
+    # losing precision on purpose due to upcasts
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    k1 = ks[0]
+    k2 = ks[1]
+    k3 = ks[2]
+    r1 = rs[0]
+    r2 = rs[1]
+
+    {_, xs, _} =
+      while {x = 0, xs, rs}, x < 4 do
+        {x + 1, apply_round(xs, rs[0][x]), rs}
+      end
+
+    xs1 = k1 + xs[0]
+
+    xs2 = k2 + i + 1 + xs[1]
+
+    new_xs =
+      Nx.stack([xs1, xs2])
+      |> Nx.as_type({:u, 32})
+
+    new_ks = Nx.stack([k2, k3, k1])
+    new_rs = Nx.stack([r2, r1])
+
+    {new_xs, new_ks, new_rs}
+  end
+
+  defnp rotate_left(x, rot) do
+    x <<< rot ||| x >>> (@nbits - rot)
+  end
+
+  defnp float_info(type) do
+    case type do
+      {:bf, 16} -> [exp: 8, mantissa: 7]
+      {:f, 16} -> [exp: 6, mantissa: 10]
+      {:f, 32} -> [exp: 8, mantissa: 23]
+      {:f, 64} -> [exp: 11, mantissa: 52]
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [min_val, max_val).
+  
+  ## Options
+  
+    * `:type` - the integer type for the returned tensor
+    * `:shape` - shape of the returned tensor
+  
+  
+  """
+  defn randint(key, min_val, max_val, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
```

Comment:
We should default to s64 which is the default for integers, WDYT?

```suggestion
    opts = keyword!(opts, shape: {1}, type: {:s, 64})
```

Reply:
Sure, adjusted.